aptos-move/framework/src/natives/cryptography/algebra/mod.rs

```diff
@@ -0,0 +1,954 @@
+            "0x1::algebra_bls12381::Gt" => Some(Structure::BLS12381Gt),
+            _ => None,
+        }
+    }
+}
+
+#[derive(Copy, Clone, Eq, Hash, PartialEq)]
+pub enum SerializationFormat {
+    BLS12381Fq12LscLscLscLsb,
+    BLS12381G1AffineCompressed,
+    BLS12381G1AffineUncompressed,
+    BLS12381G2AffineCompressed,
+    BLS12381G2AffineUnompressed,
+    BLS12381Gt,
+    BLS12381FrLsb,
+    BLS12381FrMsb,
+}
+
+impl TryFrom<u64> for SerializationFormat {
+    type Error = ();
+
+    fn try_from(value: u64) -> Result<Self, Self::Error> {
+        match value {
+            0x0400000000000000 => Ok(SerializationFormat::BLS12381Fq12LscLscLscLsb),
+            0x0600000000000000 => Ok(SerializationFormat::BLS12381G1AffineUncompressed),
+            0x0601000000000000 => Ok(SerializationFormat::BLS12381G1AffineCompressed),
+            0x0800000000000000 => Ok(SerializationFormat::BLS12381G2AffineUnompressed),
+            0x0801000000000000 => Ok(SerializationFormat::BLS12381G2AffineCompressed),
+            0x0900000000000000 => Ok(SerializationFormat::BLS12381Gt),
+            0x0A00000000000000 => Ok(SerializationFormat::BLS12381FrLsb),
+            0x0A01000000000000 => Ok(SerializationFormat::BLS12381FrMsb),
+            _ => Err(()),
+        }
+    }
+}
+
+/// Hash-to-structure suites.
+pub enum HashToStructureSuite {
+    Bls12381g1XmdSha256SswuRo,
+    Bls12381g2XmdSha256SswuRo,
+}
+
+impl TryFrom<u64> for HashToStructureSuite {
+    type Error = ();
+
+    fn try_from(value: u64) -> Result<Self, Self::Error> {
+        match value {
+            0x0001000000000000 => Ok(HashToStructureSuite::Bls12381g1XmdSha256SswuRo),
+            0x0002000000000000 => Ok(HashToStructureSuite::Bls12381g2XmdSha256SswuRo),
+            _ => Err(()),
+        }
+    }
+}
+
+#[derive(Tid, Default)]
+pub struct AlgebraContext {
+    objs: Vec<Rc<dyn Any>>,
+}
+
+impl AlgebraContext {
+    pub fn new() -> Self {
+        Self { objs: Vec::new() }
+    }
+}
+
+macro_rules! structure_from_ty_arg {
+    ($context:expr, $typ:expr) => {{
+        let type_tag = $context.type_to_type_tag($typ).unwrap();
+        Structure::from_type_tag(&type_tag)
+    }};
+}
+
+macro_rules! store_obj {
+    ($context:expr, $obj:expr) => {{
+        let target_vec = &mut $context.extensions_mut().get_mut::<AlgebraContext>().objs;
+        let ret = target_vec.len();
+        target_vec.push(Rc::new($obj));
+        ret
+    }};
+}
+
+macro_rules! abort_unless_feature_enabled {
+    ($context:ident, $feature:expr) => {
+        if !$context.get_feature_flags().is_enabled($feature) {
+            return Err(SafeNativeError::Abort {
+                abort_code: MOVE_ABORT_CODE_NOT_IMPLEMENTED,
+            });
+        }
+    };
+}
+
+macro_rules! get_obj_pointer {
+    ($context:expr, $handle:expr) => {{
+        $context.extensions_mut().get_mut::<AlgebraContext>().objs[$handle].clone()
+    }};
+}
+
+macro_rules! ark_serialize_internal {
```

Comment:
These `ark_*_internal` macros helps deduplicate code when later we use arkworks library to implement other structures.